chrysom/client.go

```diff
@@ -130,153 +211,200 @@ func validateConfig(config *ClientConfig) error {
 }
 
-func (c *Client) Remove(id string, owner string) (model.Item, error) {
-	if id == "" {
-		return model.Item{}, errors.New("id can't be empty")
-	}
-	request, err := http.NewRequest("DELETE", fmt.Sprintf("%s/api/v1/store/%s/%s", c.remoteStoreAddress, c.bucketName, id), nil)
-	if err != nil {
-		return model.Item{}, err
-	}
-	err = acquire.AddAuth(request, c.auth)
+// RemoveItem removes the item if it exists and returns the data associated to it.
+func (c *Client) RemoveItem(id, bucket, owner string) (model.Item, error) {
+	err := validateRemoveItemInput(bucket, id)
 	if err != nil {
 		return model.Item{}, err
 	}
 
-	request.Header.Add(store.ItemOwnerHeaderKey, owner)
-
-	response, err := c.client.Do(request)
+	resp, err := c.sendRequest(owner, http.MethodDelete, fmt.Sprintf("%s/%s/%s", c.storeBaseURL, bucket, id), nil)
 	if err != nil {
 		return model.Item{}, err
 	}
-	if response.StatusCode != 200 {
-		return model.Item{}, errors.New("failed to delete item, non 200 statuscode")
+
+	if resp.Code != http.StatusOK {
+		level.Error(c.logger).Log(xlog.MessageKey(), "Argus responded with a non-successful status code for a RemoveItem request",
+			"code", resp.Code, "ErrorHeader", resp.ArgusErrorHeader)
+		return model.Item{}, fmt.Errorf("statusCode %v: %w", resp.Code, translateNonSuccessStatusCode(resp.Code))
 	}
-	defer response.Body.Close()
-	responsePayload, _ := ioutil.ReadAll(response.Body)
-	item := model.Item{}
-	err = json.Unmarshal(responsePayload, &item)
+
+	var item model.Item
+	err = json.Unmarshal(resp.Body, &item)
 	if err != nil {
-		return model.Item{}, err
+		return item, fmt.Errorf("RemoveItem: %w: %s", errJSONUnmarshal, err.Error())
 	}
 	return item, nil
 }
 
 func (c *Client) Start(ctx context.Context) error {
-	if c.ticker == nil {
-		return errors.New("interval ticker is nil")
+	if c.observer == nil {
+		level.Warn(c.logger).Log(xlog.MessageKey(), "No listener was setup to receive updates.")
+		return nil
 	}
 
-	if c.listener == nil {
-		level.Info(c.logger).Log(xlog.MessageKey(), "No listener was setup to receive updates.")
-		return nil
+	if c.observer.ticker == nil {
+		return ErrUndefinedIntervalTicker
 	}
 
 	go func() {
-		for range c.ticker.C {
+		observer := c.observer
+		for range observer.ticker.C {
 			outcome := SuccessOutcome
-			items, err := c.GetItems("")
+			items, err := c.GetItems(observer.bucket, observer.owner)
 			if err == nil {
-				c.listener.Update(items)
+				observer.listener.Update(items)
 			} else {
 				outcome = FailureOutcome
 				level.Error(c.logger).Log(xlog.MessageKey(), "Failed to get items for listeners", xlog.ErrorKey(), err)
 			}
-			c.metrics.pollCount.With(OutcomeLabel, outcome).Add(1)
+			observer.pollCount.With(OutcomeLabel, outcome).Add(1)
 		}
 	}()
 
 	return nil
 }
 
 func (c *Client) Stop(ctx context.Context) error {
-	if c.ticker != nil {
-		c.ticker.Stop()
+	if c.observer != nil && c.observer.ticker != nil {
+		c.observer.ticker.Stop()
+	}
```

Comment:
This doesn't need to be done in this pr, but should we add a `WaitGroup` to ensure that the long running go routine has exited before returning? if `GetItems()` takes time but is successful, we could be trying to update the listener after it is already shut down.

Reply:
Yeah, I think that's a great idea to avoid things sneakily happening in the background after Stop() returns 🚀 
I'll add an issue for it